firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,544 @@
+
+        url = self.marionette.execute_script("""
+            Cu.import("resource://gre/modules/Services.jsm");
+
+            var url = arguments[0];
+            return Services.urlFormatter.formatURL(url);
+
+            """, script_args=[url])
+
+        if force:
+            if '?' in url:
+                url += '&'
+            else:
+                url += '?'
+            url += 'force=1'
+        return url
+
+
+class ActiveUpdate(object):
+
+    def __init__(self, software_update):
+
+        self.marionette = software_update.marionette
+
+    def __getattr__(self, attr):
+        value = self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            var active_update = ums.activeUpdate;
+            return active_update[arguments[0]];
+        """, script_args=[attr])
+
+        if value is not None:
+            return value
+        else:
+            raise AttributeError('{} has no attribute {}'.format(self.__class__.__name__,
+                                                                 attr))
+
+    @property
+    def exists(self):
+        """Is there an active update
+
+        :returns: True if there is an active update
+        """
+
+        active_update = self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            return ums.activeUpdate;
+
+        """)
+
+        return bool(active_update)
+
+    def get_patch_at(self, patch_index):
+        """Use nsIUpdate.getPatchAt to return a patch from an update
+
+        :returns: nsIUpdatePatch object
+        """
+
+        return self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            let active_update = ums.activeUpdate;
+            let patch_index = arguments[0];
+            return active_update.getPatchAt(patch_index);
+
+        """, script_args=[patch_index])
+
+    @property
+    def patch_count(self):
+        """Get the patchCount from the active update
+
+        :returns: The patch count
+        """
+
+        return self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            let active_update = ums.activeUpdate;
+            return active_update.patchCount;
+
+        """)
+
+    @property
+    def selected_patch_final_url(self):
+        """Get the finalURL of the selected patch for the active update
+
+        :returns: The patch size
+        """
+
+        return self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            let active_update = ums.activeUpdate;
+            return active_update.selectedPatch.finalURL;
```

Comment:
Hm, maybe we can also have a `selected_patch` property only and return the JSON object? That way we have only a single method and using that property will become easier.